source/common/thread_local/thread_local_impl.cc

```diff
@@ -94,6 +94,8 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb) {
 void InstanceImpl::runOnAllThreads(Event::PostCb cb, Event::PostCb all_threads_complete_cb) {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
   ASSERT(!shutdown_);
+  // Handle main thread.
+  cb();
```

Comment:
in the runOnAllThreads variant above, the main thread happens last.  Here you are running on the main thread concurrently.  Doing that here will be slightly more difficult because you'l have to incorporate that into your worker_count decrement semantics.  So if you are not going to do that, it might be worth some commentary about why it needs to be different.